[Ernestnzx] iP

Ernestnzx opened the pull request:
# Amara Chatbot 💙 
> 'Insanity is doing the same thing over and over again and expecting different results.' - Albert Einstein

Instead of trying to remember things over and over again and letting it stick in your memory, try Amara Bot! 
- It is command / text-based
- Remembers and keeps track of all of your tasks
- ~~Fast~~ **SUPER FAST** to use!

Why wait any longer? Here are the steps to use AmaraBot✨
1. Download the application [here](https://github.com/Ernestnzx/ip)
2. Double click on the `jar` file
3. Save your tasks to the list
4. ???
5. Profit

And it's **FREE!!**

### Features:
- [X] Todo, Event, Deadline tasks implemented
- [ ] Cleanup outdated tasks (next feature)

If you are a **Java** programmer and you like ~~trauma~~ coding practices here is an exercise!
```java
public static void main(String[] args) {
   System.out.println("Hello I'm Amara\nWhat can I do for you?");
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Ernestnzx committed:
Level-0

Ernestnzx committed:
Added echo

Ernestnzx committed:
Added task lists

Ernestnzx committed:
Added Task + changed bot reply in Main

Ernestnzx committed:
Fixed spelling mistake

Ernestnzx committed:
Addded Task interface + refactor code

Ernestnzx committed:
Changed interface to abstract class

Ernestnzx committed:
Added all events + fixed .toString() methods

Ernestnzx committed:
Added TaskFactory + Events

Ernestnzx committed:
Added Automated Text UI Testing

Ernestnzx committed:
Moved logic from Main to Amara Bot

Ernestnzx committed:
Added error testcase + code for exception handling

Ernestnzx committed:
Added more exception handling

Ernestnzx committed:
Added delete functionality

Ernestnzx committed:
Swapped if-else to switch-case

Ernestnzx committed:
Added enums and changed the switch-case

Ernestnzx committed:
Changed all instances of commandString to it's enum

Ernestnzx committed:
Added save functionality

Ernestnzx committed:
Merge branch 'branch-Level-7'

Ernestnzx committed:
Amara bot now understands date and time

Ernestnzx committed:
Merge branch 'branch-Level-8'

Ernestnzx committed:
Added Storage + Ui for AmaraBot

Ernestnzx committed:
Updated UI class to print messages

Ernestnzx committed:
Fixed Memory file issue

Ernestnzx committed:
Refactored my spaghetti to make it more OOP

Ernestnzx committed:
Merge branch 'branch-A-MoreOOP'

Ernestnzx committed:
Added packages to make better spaghetti

Ernestnzx committed:
Merge branch 'branch-A-Packages'

Ernestnzx committed:
Removed unused code

Ernestnzx committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Ernestnzx committed:
Amara is now the main class

Ernestnzx committed:
Merged with gradle and updated config

Ernestnzx committed:
Add JUnit Tests

Ernestnzx committed:
Close BufferedReader

Ernestnzx committed:
Change archiveBaseName to amara

Ernestnzx committed:
Added Java documentation for all classes

Ernestnzx committed:
Align the code better to codestyle

Ernestnzx committed:
Add Find functionality

Ernestnzx committed:
Merge branch 'branch-A-JavaDoc'

Ernestnzx committed:
Merge branch 'branch-A-CodingStandard'

Ernestnzx committed:
Merge branch 'branch-Level-9'

Ernestnzx committed:
Add data + fix Memory bug

Ernestnzx committed:
Fix Memory bug, rearrange and add more test cases

Ernestnzx committed:
Change return signature of Command from void to String

Ernestnzx committed:
Fix indent for switch-case

Ernestnzx committed:
Update README.md

Ernestnzx committed:
Add checkstyle and Javadocs

Ernestnzx committed:
Merge branch 'branch-A-Checkstyle'

Ernestnzx committed:
Add GUI to Amara bot

Ernestnzx committed:
Add help command

Ernestnzx committed:
Merge branch 'branch-Level-10'

Ernestnzx committed:
Change CSS style + Update FindCommand

Ernestnzx committed:
Move javafx classes to ui folder

Ernestnzx committed:
Add config files

Ernestnzx committed:
Edit gitignore for setting files

vrisdng commented on src/main/java/AmaraException.java:
```diff
@@ -1,7 +1,8 @@
     private static final String ITS_THE_END = "OWARI DAAAAA!!!!!!";
     private static final String ERROR_MESSAGE = "OOPS!!! I'm sorry, but I don't know what that means. :(";
-    private static final String OUT_OF_BOUNDS = "OOPS!!! The index you want to mark/unmark is out of bounds :(";
+    private static final String OUT_OF_BOUNDS = "OOPS!!! The index you want to mark/unmark is out of bounds. :(";
```
Wow i like the way you write the exception messages, kudos!

vrisdng commented on src/main/java/amara/command/Parser.java:
```diff
@@ -0,0 +1,117 @@
+     * @return A {@link Command} object that corresponds to the parsed command.
+     * @throws AmaraException If the command is invalid or if an error occurs while parsing.
+     */
+    public static Command parseCommand(String fullCommand) throws AmaraException {
```
Nice implementation, very clean

vrisdng commented on src/main/java/amara/command/Parser.java:
```diff
@@ -0,0 +1,117 @@
+            if (tokens.length != 2 || tokens[0].isBlank()) {
+                throw AmaraException.invalidDeadlineParameter();
+            }
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
```
Do you think if it is better to extract the date-time from the input separately, then for this method you only have to parse what you got into the format you want?

delonlee01 commented on src/main/java/amara/command/AddEventCommand.java:
```diff
@@ -0,0 +1,40 @@
+import amara.ui.Ui;
+
+/**
+ * A {@link Command} implementation that adds a {@link Deadline} task
```
```suggestion
 * A {@link Command} implementation that adds a {@link Event} task
```

delonlee01 commented on src/main/java/amara/command/AddToDoCommand.java:
```diff
@@ -0,0 +1,40 @@
+import amara.ui.Ui;
+
+/**
+ * A {@link Command} implementation that adds a {@link Deadline} task
```
```suggestion
 * A {@link Command} implementation that adds a {@link ToDo} task
```

delonlee01 commented on src/main/java/amara/command/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+    public String execute(ArrayList<Task> tasks, Ui ui, Storage storage) throws AmaraException {
+        try {
+            Task task = tasks.get(this.index);
+            tasks.remove(this.index);
```
IIRC, `ArrayList::remove` returns the element removed. Maybe you can consider combining these two operations?

delonlee01 commented on src/main/java/amara/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,44 @@
+import amara.ui.Ui;
+
+/**
+ * A {@link Command} that unmarks a {@link Task} as incomplete
```
```suggestion
 * A {@link Command} that marks a {@link Task} as incomplete
```

delonlee01 commented on src/main/java/amara/ui/Amara.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    public String greet() {
```
Maybe `getGreeting` is more appropriate?

delonlee01 reviewed (commented):
Accident. Please ignore.

Ernestnzx committed:
Add assertions

Ernestnzx committed:
Improve code quality

Ernestnzx committed:
Add streams to find and list command

Ernestnzx committed:
Merge branch 'branch-A-Assertions'

Ernestnzx committed:
Resolve merge conflict with master

Ernestnzx committed:
Merge branch 'master' into branch-A-Streams

Ernestnzx committed:
Resolve merge conflict with master

Ernestnzx committed:
Made profile image a circle

potatodudedude commented on src/main/java/amara/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,39 @@
+        this.deadline = deadline;
+    }
+
+    /**
+     * Executes the {@code Command}.
+     * @param tasks List of tasks.
+     * @param ui UI handler.
+     * @param storage To store the given List of tasks.
+     */
```
```suggestion
    /**
     * Adds the Deadline to a given ArrayList<Task>, forms a message for a UI handler to display
     * 
     * @param tasks List of tasks.
     * @param ui UI handler.
     * @param storage To store the given List of tasks.
     * @return Message given to UI handler
     */
```
Coding standard stuff: A line is needed between the description and the parameters. I think all your method JavaDocs have this problem.
Also, the description is quite vague, perhaps you can consider elaborating a little more, something like the example given.

potatodudedude commented on src/main/java/amara/command/AddToDoCommand.java:
```diff
@@ -0,0 +1,39 @@
+public class AddToDoCommand extends Command {
+    private static final String MESSAGE = "Got it. I've added this task:\n"
+            + "  %s\nNow you have %d tasks in the list.";
+    private ToDo todo;
```
```suggestion
    private ToDo toDo;
```
Coding standard stuff: variable name should be camelCase

potatodudedude commented on src/main/java/amara/command/DeleteCommand.java:
```diff
@@ -0,0 +1,44 @@
+public class DeleteCommand extends Command {
+    private static final String MESSAGE = "Noted. I've removed this task:\n"
+            + "  %s\nNow you have %d tasks in the list.";
+    private final int index;
```
Perhaps you could make this variable name more specific? e.g. taskIndex
Same situation for your MarkCommand and UnmarkCommand classes.

potatodudedude commented on src/main/java/amara/task/Task.java:
```diff
@@ -0,0 +1,44 @@
+ */
+public abstract class Task {
+    protected String taskDescription;
+    protected boolean status;
```
I think you can make the boolean naming clearer, like isDone or isMarked.

potatodudedude commented on src/main/java/amara/ui/Amara.java:
```diff
@@ -0,0 +1,85 @@
+public class Amara {
+    private static final String FILE_PATH = System.getProperty("user.dir") + "/data/taskfile.txt";
+    private static final String GREETING_MESSAGE = "Hello I'm Amara\nWhat can I do for you?";
+    private ArrayList<Task> tasks;
```
Why did you decide to not use a class to represent the task list?

potatodudedude reviewed (commented):
Your code is very well put together! Just a few small things here and there. Keep it up 👍

Ernestnzx committed:
Edit Javadocs to match the standard.

Ernestnzx committed:
Edit Javadocs to match required checkstyle

Ernestnzx committed:
Add Sort functionality

Ernestnzx committed:
Add Task comparator to sort task with LocalDateTime

Deadline and Event tasks are originally sorted lexicographically
by their task description. Now they are sorted by due dute and
start date respectively and ties are broken by their description.

Ernestnzx committed:
Merge branch 'branch-C-Sort'

Ernestnzx committed:
Update help command and greeting message

Update list command to describe the funcionality of calling
the sort method on the given task list. Change the greeting
message of Amara the eevee bot to give it more personality.

Ernestnzx committed:
Add markdown for list of AI assistance

Ernestnzx committed:
Merge pull request #2 from Ernestnzx/branch-AiAssisted

Add markdown for list of AI assistance

Ernestnzx committed:
Enable Amara to save taskList on every command

Ernestnzx committed:
Add Ui.png

Ernestnzx committed:
Update user manual

Ernestnzx committed:
Edit markdown for formatting

Ernestnzx committed:
Fix formatting issues for user manual

Ernestnzx committed:
Update README.md

Ernestnzx committed:
Edit find method description

Ernestnzx committed:
Update README.md

Ernestnzx committed:
Update README.md

Ernestnzx committed:
Fix grammar and formatting issues
